server/src/main/java/io/littlehorse/server/streams/stores/ReadOnlyBaseStoreImpl.java

```diff
@@ -0,0 +1,123 @@
+package io.littlehorse.server.streams.stores;
+
+import com.google.protobuf.Message;
+import io.littlehorse.common.LHSerializable;
+import io.littlehorse.common.Storeable;
+import io.littlehorse.sdk.common.exception.LHSerdeError;
+import io.littlehorse.server.streams.store.LHKeyValueIterator;
+import io.littlehorse.server.streams.topology.core.ExecutionContext;
+import java.util.Optional;
+import lombok.Getter;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.state.ReadOnlyKeyValueStore;
+
+/**
+ * Package-private class used to implement multiple logical stores within a single Kafka Streams
+ * KeyValueStore. This is motivated purely for performance reasons because adding multiple stores
+ * leads to a proliferation of open files (RocksDB), many changelog partitions, slower rebalances,
+ * more partitions per Streams Transaction (longer commits), and slower flushes.
+ *
+ * In addition to allowing different Storeable's to be stored in logically separated stores, this
+ * class runs in two modes: Cluster-scoped, or Tenant-scoped. In the Tenant-Scoped mode, a prefix
+ * is pre-pended to every key so that we can logically isolate Tenant-Scoped data.
+ */
+abstract class ReadOnlyBaseStoreImpl implements ReadOnlyBaseStore {
+
+    @Getter
+    protected final String tenantId;
+
+    protected final ExecutionContext executionContext;
+    private final ReadOnlyKeyValueStore<String, Bytes> nativeStore;
+
+    ReadOnlyBaseStoreImpl(
```

Comment:
This is not a good way to represent optional arguments for constructors. Instead, you should create two constructors:

- `ReadOnlyBaseStoreImpl(ReadOnlyKeyValueStore<String, Bytes> nativeStore, ExecutionContext executionContext)`
- `ReadOnlyBaseStoreImpl(ReadOnlyKeyValueStore<String, Bytes> nativeStore,String tenantId,ExecutionContext executionContext)`

 Optional is not intended to be a replacement for every null reference. 

Sames applies for specific classes of these stores.

Reply:
Also, I propose reducing the access modifier for these constructors to protected.